sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala

```diff
@@ -289,48 +289,90 @@ object AlluxioUtils extends Logging {
-      val inputFiles: Seq[Path] = partitionDirs.flatMap(partitionDir => {
-        partitionDir.files.map(f => replaceFunc.get(f.getPath))
-      })
+      def replacePathsInPartitionSpec(spec: PartitionSpec): PartitionSpec = {
+        val partitionsWithPathsReplaced = spec.partitions.map { p =>
+          val replacedPath = replaceFunc.get(p.path)
+          org.apache.spark.sql.execution.datasources.PartitionPath(p.values, replacedPath)
+        }
+        PartitionSpec(spec.partitionColumns, partitionsWithPathsReplaced)
+      }
 
-      // replace all of rootPaths which are already unique
-      val rootPaths = relation.location.rootPaths.map(replaceFunc.get)
+      def createNewFileIndexWithPathsReplaced(
+          spec: PartitionSpec,
+          rootPaths: Seq[Path]): InMemoryFileIndex = {
+        val specAdjusted = replacePathsInPartitionSpec(spec)
+        val replacedPaths = rootPaths.map(p => replaceFunc.get(p))
+        new InMemoryFileIndex(
+          relation.sparkSession,
+          replacedPaths,
+          relation.options,
+          Option(relation.dataSchema),
+          userSpecifiedPartitionSpec = Some(specAdjusted))
+      }
 
-      // check the alluxio paths in root paths exist or not
-      // throw out an exception to stop the job when any of them is not mounted
-      if (replaceMapOption.isDefined) {
-        rootPaths.foreach { rootPath =>
-          replaceMapOption.get.values.find(value => rootPath.toString.startsWith(value)).foreach {
-            matched => checkAlluxioMounted(relation.sparkSession, matched)
+      // If we know the type of file index, try to reuse as much of the existing
+      // FileIndex as we can to keep from having to recalculate it and potentially
+      // mess it up. Things like partitioning are already included and some of the
+      // Spark infer code didn't handle certain types properly. Here we try to just
+      // update the paths to the new Alluxio path. If its not a type of file index
+      // we know then fall back to inferring. The latter happens on certain CSPs
+      // like Databricks where they have customer file index types.
+      if (relation.location.isInstanceOf[PartitioningAwareFileIndex]) {
+        logDebug("Handling PartitioningAwareFileIndex")
+        val pfi = relation.location.asInstanceOf[PartitioningAwareFileIndex]
+        createNewFileIndexWithPathsReplaced(pfi.partitionSpec(), pfi.rootPaths)
+      } else if (relation.location.isInstanceOf[CatalogFileIndex]) {
+        logDebug("Handling CatalogFileIndex")
+        val cfi = relation.location.asInstanceOf[CatalogFileIndex]
+        val memFI = cfi.filterPartitions(Nil)
+        createNewFileIndexWithPathsReplaced(memFI.partitionSpec(), memFI.rootPaths)
+      } else {
+        logDebug(s"Handling file index type: ${relation.location.getClass}")
+        // With the base Spark FileIndex type we don't know how to modify it to
+        // just replace the paths so we have to try to recompute.
```

Comment:
is there a finite list of FileIndex types that we would handle here? E.g. could we refer to them so we know why we are not special casing them.

Reply:
I think the answer is yes... any datasource could create their own fileIndex type... like for delta for instance. If its not a type we know then we don't know what api they might be using to take into account partitioning and things like that.